src/gt4py/next/utils.py

```diff
@@ -69,26 +69,76 @@ def flatten_nested_tuple(value: tuple[_T | tuple, ...]) -> tuple[_T, ...]:
         return (value,)
 
 
-def tree_map(fun: Callable[_P, _R]) -> Callable[..., _R | tuple[_R | tuple, ...]]:
+@overload
+def tree_map(fun: Callable[_P, _R], /) -> Callable[..., _R | tuple[_R | tuple, ...]]: ...
+
+
+@overload
+def tree_map(
+    *, collection_type: type | tuple[type, ...], result_collection_type: Optional[type] = None
+) -> Callable[[Callable[_P, _R]], Callable[..., _R | tuple[_R | tuple, ...]]]: ...
+
+
+def tree_map(
```

Comment:
I think it is time to remove `apply_to_primitive_constituents` and absorb it in this function. I'll check if @SF-N has capacity to work on this.

Reply:
Sara will extract this into a new PR and also remove the `apply_to_primitive_constituents` cases there. I'll keep you in the loop.